### DISCUSSION TITLE
How to initialize trainable_lemmatizer in config.cfg?

### DISCUSSION THREAD

**Discussion by canberkarc** [explosion/spaCy#11167] [Help: Coding & Implementations] [answered]
When I run this command:

python -m spacy train "C:/Users/canberk/Desktop/Model/config.cfg" --output "C:/Users/canberk/Desktop/Model/models/output"

I'm getting the error below:

" ValueError: [E143] Labels for component 'trainable_lemmatizer' not initialized. This can be fixed by calling add_label, or by providing a representative batch of examples to the component's `initialize` method. "

Here is config.cfg:

```

[paths]
train = "C:/Users/canberk/Desktop/ Model/data/train.spacy"
dev = "C:/Users/canberk/Desktop/ Model/data/valid.spacy"
vectors = null
init_tok2vec = null

[system]
gpu_allocator = null
seed = 0

[nlp]
lang = "tr"
pipeline = ["tok2vec","parser","trainable_lemmatizer","ner"]
batch_size = 1000
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.ner]
factory = "ner"
incorrect_spans_key = null
moves = null
scorer = {"@scorers":"spacy.ner_scorer.v1"}
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = true
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.parser]
factory = "parser"
learn_tokens = false
min_action_freq = 30
moves = null
scorer = {"@scorers":"spacy.parser_scorer.v1"}
update_with_oracle_cut_size = 100

[components.parser.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "parser"
extra_state_tokens = false
hidden_width = 128
maxout_pieces = 3
use_upper = true
nO = null

[components.parser.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[components.tok2vec]
factory = "tok2vec"

[components.tok2vec.model]
@architectures = "spacy.Tok2Vec.v2"

[components.tok2vec.model.embed]
@architectures = "spacy.MultiHashEmbed.v2"
width = ${components.tok2vec.model.encode.width}
attrs = ["NORM","PREFIX","SUFFIX","SHAPE"]
rows = [5000,2500,2500,2500]
include_static_vectors = false

[components.tok2vec.model.encode]
@architectures = "spacy.MaxoutWindowEncoder.v2"
width = 96
depth = 4
window_size = 1
maxout_pieces = 3

[components.trainable_lemmatizer]
factory = "trainable_lemmatizer"
backoff = "orth"
min_tree_freq = 3
overwrite = false
scorer = {"@scorers":"spacy.lemmatizer_scorer.v1"}
top_k = 1

[components.trainable_lemmatizer.model]
@architectures = "spacy.Tagger.v2"
nO = null
normalize = false

[components.trainable_lemmatizer.model.tok2vec]
@architectures = "spacy.Tok2VecListener.v1"
width = ${components.tok2vec.model.encode.width}
upstream = "*"

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
annotating_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_words.v1"
discard_oversize = false
tolerance = 0.2
get_length = null

[training.batcher.size]
@schedules = "compounding.v1"
start = 100
stop = 1000
compound = 1.001
t = 0.0

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = false

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
dep_uas = 0.17
dep_las = 0.17
dep_las_per_type = null
sents_p = null
sents_r = null
sents_f = 0.0
lemma_acc = 0.33
ents_f = 0.33
ents_p = 0.0
ents_r = 0.0
ents_per_type = null

[pretraining]

[initialize]
vectors = ${paths.vectors}
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]

```

**Comment by polm** [comment, marked as answer]
To format code blocks correctly, you need to put three backticks at the start and end. See the [Github Markdown guide](https://docs.github.com/en/get-started/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax#quoting-code). 

In order for the trainable component to work, you need to provide labels. You can do this by providing a file with labels or annotated data. For the lemmatizer it's much easier to provide annotated data.

This error suggests that your training data doesn't have any lemmas assigned. How are you creating your training data?

**Reply by canberkarc** [reply]
Sorry for inconvenience. I didn't prepare a data for lemmas but I can prepare. How can I provide this data to config.cfg or trainable_lemmatizer itself?

**Reply by polm** [reply]
To clarify, do you actually want to train a lemmatizer, or do you just want to use the lemmatizer from the pretrained pipeline? You don't need to train a lemmatizer unless you don't like the way the existing one works or you're using a new language.

If you do want to train a lemmatizer, you need to set lemmas on a training document. That would work similarly to the sample code in the [training data docs](https://spacy.io/usage/training/#training-data), except that it would be a little like this:

```
for tok in doc:
    tok.lemma_ = "lemma" # in reality this should come from a list or something
```

You could also use `Doc(nlp.vocab, words=[...], lemmas=[...])` to make a Doc, see the [API documentation](https://spacy.io/api/doc#init).

Could you explain what your overall project or goal is? That might make it easier for us to help you.

**Reply by canberkarc** [reply]
Actually I have installed spacy lookup tables from https://pypi.org/project/spacy-lookups-data/ and I can reach this table for Turkish if I create pipeline on jupyter notebook for example, by writing ``` nlp.get_pipe("lemmatizer").lookups.get_table("lemma_lookup")```.
I want to include lemmatizer in pipeline in config file but I guess it doesn't reach installed spacy lookup tables, I couldn't figure out what is the problem.

**Reply by polm** [reply]
To add the lemmatizer to your pipeline you need to add it to the `pipeline` in the `[nlp]` section and then add a section like this:

```
[components.lemmatizer]
factory = "lemmatizer"
mode = "lookup"
```

The lookup tables are for the standard lemmatizer, not the trainable lemmatizer. The trainable lemmatizer uses a statistical model instead of lookup tables, which is why it needed annotated data.
